src/size_hint.rs

```diff
@@ -1,149 +1,493 @@
-//! Utilities for working with and combining the results of
-//! [`Arbitrary::size_hint`][crate::Arbitrary::size_hint].
+//! The [`SizeHint`] type, for use with
+//! [`Arbitrary::size_hint()`].
 
-pub(crate) const MAX_DEPTH: usize = 20;
+use core::cell::Cell;
+use core::{fmt, ops};
 
-/// Protects against potential infinite recursion when calculating size hints
-/// due to indirect type recursion.
-///
-/// When the depth is not too deep, calls `f` with `depth + 1` to calculate the
-/// size hint.
-///
-/// Otherwise, returns the default size hint: `(0, None)`.
-///
-/// <div class="warning">This method is deprecated. Users should instead implement <a href="../trait.Arbitrary.html#method.try_size_hint"><code>try_size_hint</code></a> and use <a href="fn.try_recursion_guard.html"><code>try_recursion_guard</code></a></div>
-#[inline]
-#[deprecated(note = "use `try_recursion_guard` instead")]
-pub fn recursion_guard(
-    depth: usize,
-    f: impl FnOnce(usize) -> (usize, Option<usize>),
-) -> (usize, Option<usize>) {
-    if depth > MAX_DEPTH {
-        (0, None)
-    } else {
-        f(depth + 1)
-    }
-}
-
-/// Protects against potential infinite recursion when calculating size hints
-/// due to indirect type recursion.
-///
-/// When the depth is not too deep, calls `f` with `depth + 1` to calculate the
-/// size hint.
-///
-/// Otherwise, returns an error.
+use crate::Arbitrary;
+
+/// Bounds on the number of bytes an [`Arbitrary::arbitrary()`] implementation might consume.
 ///
-/// This should be used when implementing [`try_size_hint`](crate::Arbitrary::try_size_hint)
-#[inline]
-pub fn try_recursion_guard(
-    depth: usize,
-    f: impl FnOnce(usize) -> Result<(usize, Option<usize>), crate::MaxRecursionReached>,
-) -> Result<(usize, Option<usize>), crate::MaxRecursionReached> {
-    if depth > MAX_DEPTH {
-        Err(crate::MaxRecursionReached {})
-    } else {
-        f(depth + 1)
+/// A size hint consists of a finite lower bound and a possibly infinite upper bound.
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+pub struct SizeHint {
+    /// Inclusive lower bound.
+    lower: usize,
+
+    /// Exclusive upper bound. [`None`] means unbounded.
+    ///
+    /// Also contains some meta-information about the size hint as a whole.
+    upper: Upper,
+
+    out_of_fuel: bool,
```

Comment:
I think we will want a method on `SizeHint`, and probably `Context` too, for checking whether we are out of fuel. I didn't see these below; apologies if these exist and I just missed it in my skim.